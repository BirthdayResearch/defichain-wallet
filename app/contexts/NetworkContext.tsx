import { NetworkName } from '@defichain/jellyfish-network'
import React, { createContext, useContext, useEffect, useState } from 'react'
import { Logging } from '../api/logging'
import * as storage from '../api/storage'
import { EnvironmentNetwork } from '../environment'

interface Network {
  network: EnvironmentNetwork
  networkName: NetworkName
  updateNetwork: (network: EnvironmentNetwork) => Promise<void>
  reloadNetwork: () => Promise<void>
}

const NetworkContext = createContext<Network>(undefined as any)

function networkMapper (network: EnvironmentNetwork): NetworkName {
  switch (network) {
    case EnvironmentNetwork.MainNet:
      return 'mainnet'
    case EnvironmentNetwork.TestNet:
      return 'testnet'
    case EnvironmentNetwork.LocalPlayground:
    case EnvironmentNetwork.RemotePlayground:
      return 'regtest'
  }
}

export function useNetworkContext (): Network {
  return useContext(NetworkContext)
}

export function NetworkProvider (props: React.PropsWithChildren<any>): JSX.Element | null {
  const [network, setNetwork] = useState<EnvironmentNetwork | undefined>(undefined)

  useEffect(() => {
    storage.getNetwork().then(async value => {
      setNetwork(value)
    }).catch(Logging.error)
  }, [])

  if (network === undefined) {
    return null
  }

  const context: Network = {
    network: network,
    networkName: networkMapper(network),
    async updateNetwork (value: EnvironmentNetwork): Promise<void> {
      await storage.setNetwork(value)
      setNetwork(value)
    },
    async reloadNetwork (): Promise<void> {
      setNetwork(network)
    }
  }

  return (
    <NetworkContext.Provider value={context}>
      {props.children}
    </NetworkContext.Provider>
  )
}
