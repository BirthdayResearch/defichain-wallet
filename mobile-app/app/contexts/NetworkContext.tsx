import { NetworkName } from '@defichain/jellyfish-network'
import React, { createContext, useContext, useEffect, useState } from 'react'
import { Logging, SecuredStoreAPI } from '@api'
import { getJellyfishNetwork } from '@api/wallet'
import { EnvironmentNetwork } from '@environment'

interface NetworkContextI {
  network: EnvironmentNetwork
  networkName: NetworkName
  updateNetwork: (network: EnvironmentNetwork) => Promise<void>
}

const NetworkContext = createContext<NetworkContextI>(undefined as any)

export function useNetworkContext (): NetworkContextI {
  return useContext(NetworkContext)
}

export function NetworkProvider (props: React.PropsWithChildren<any>): JSX.Element | null {
  const [network, setNetwork] = useState<EnvironmentNetwork>()
  const [networkName, setNetworkName] = useState<NetworkName>()

  useEffect(() => {
    SecuredStoreAPI.getNetwork().then(async value => {
      setNetworkName(getJellyfishNetwork(value).name)
      setNetwork(value)
    }).catch(Logging.error)
  }, [])

  if (network === undefined || networkName === undefined) {
    return null
  }

  const context: NetworkContextI = {
    network: network,
    networkName: networkName,
    async updateNetwork (value: EnvironmentNetwork): Promise<void> {
      await SecuredStoreAPI.setNetwork(value)
      setNetworkName(getJellyfishNetwork(value).name)
      setNetwork(value)
    }
  }

  return (
    <NetworkContext.Provider value={context}>
      {props.children}
    </NetworkContext.Provider>
  )
}
