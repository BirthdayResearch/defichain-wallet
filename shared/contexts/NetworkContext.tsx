import { NetworkName } from '@defichain/jellyfish-network'
import React, { createContext, useContext, useEffect, useState, PropsWithChildren } from 'react'
import { getJellyfishNetwork } from '@shared-api/wallet/network'
import { EnvironmentNetwork } from '@environment'
import { useLogger } from '@shared-contexts/NativeLoggingProvider'

interface NetworkContextI {
  network: EnvironmentNetwork
  networkName: NetworkName
  updateNetwork: (network: EnvironmentNetwork) => Promise<void>
}

const NetworkContext = createContext<NetworkContextI>(undefined as any)

export function useNetworkContext (): NetworkContextI {
  return useContext(NetworkContext)
}

export interface NetworkProviderProps extends PropsWithChildren<{}> {
  api: {
    getNetwork: () => Promise<EnvironmentNetwork>
    setNetwork: (network: EnvironmentNetwork) => Promise<void>
  }
}

export function NetworkProvider (props: NetworkProviderProps): JSX.Element | null {
  const [network, setNetwork] = useState<EnvironmentNetwork>()
  const [networkName, setNetworkName] = useState<NetworkName>()
  const logger = useLogger()
  const { api } = props

  useEffect(() => {
    api.getNetwork().then(async (value: EnvironmentNetwork) => {
      setNetworkName(getJellyfishNetwork(value).name)
      setNetwork(value)
    }).catch(logger.error)
  }, [])

  if (network === undefined || networkName === undefined) {
    return null
  }

  const context: NetworkContextI = {
    network: network,
    networkName: networkName,
    async updateNetwork (value: EnvironmentNetwork): Promise<void> {
      await api.setNetwork(value)
      setNetworkName(getJellyfishNetwork(value).name)
      setNetwork(value)
    }
  }

  return (
    <NetworkContext.Provider value={context}>
      {props.children}
    </NetworkContext.Provider>
  )
}
