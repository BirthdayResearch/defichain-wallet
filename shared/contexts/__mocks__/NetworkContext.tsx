export const useNetworkContext = (): { network: string } => {
  return {
    network: 'Playground'
  }
}
