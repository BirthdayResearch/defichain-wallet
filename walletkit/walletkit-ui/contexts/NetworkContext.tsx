import { NetworkName } from "@defichain/jellyfish-network";
import {
  EnvironmentNetwork,
  getJellyfishNetwork,
} from "@waveshq/walletkit-core";
import React, {
  createContext,
  PropsWithChildren,
  useContext,
  useEffect,
  useState,
} from "react";

import { BaseLogger } from "./logger";

interface NetworkContextI {
  network: EnvironmentNetwork;
  networkName: NetworkName;
  updateNetwork: (network: EnvironmentNetwork) => Promise<void>;
}

const NetworkContext = createContext<NetworkContextI>(undefined as any);

export function useNetworkContext(): NetworkContextI {
  return useContext(NetworkContext);
}

export interface NetworkProviderProps extends PropsWithChildren<{}> {
  api: {
    getNetwork: () => Promise<EnvironmentNetwork>;
    setNetwork: (network: EnvironmentNetwork) => Promise<void>;
  };
  logger: BaseLogger;
}

export function NetworkProvider(
  props: NetworkProviderProps,
): JSX.Element | null {
  const [network, setNetwork] = useState<EnvironmentNetwork>();
  const [networkName, setNetworkName] = useState<NetworkName>();
  const { api, logger, children } = props;

  useEffect(() => {
    api
      .getNetwork()
      .then(async (value: EnvironmentNetwork) => {
        setNetworkName(getJellyfishNetwork(value).name);
        setNetwork(value);
      })
      .catch(logger.error);
  }, []);

  if (network === undefined || networkName === undefined) {
    return null;
  }

  // eslint-disable-next-line react/jsx-no-constructed-context-values
  const context: NetworkContextI = {
    network,
    networkName,
    async updateNetwork(value: EnvironmentNetwork): Promise<void> {
      await api.setNetwork(value);
      setNetworkName(getJellyfishNetwork(value).name);
      setNetwork(value);
    },
  };

  return (
    <NetworkContext.Provider value={context}>
      {children}
    </NetworkContext.Provider>
  );
}
