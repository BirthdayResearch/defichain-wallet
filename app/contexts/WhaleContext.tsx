import { WhaleApiClient } from '@defichain/whale-api-client'
import React, { createContext, useContext, useMemo } from 'react'
import { EnvironmentNetwork } from '../environment'
import { networkMapper, useNetworkContext } from './NetworkContext'

const WhaleApiClientContext = createContext<WhaleApiClient>(undefined as any)

export function useWhaleApiClient (): WhaleApiClient {
  return useContext(WhaleApiClientContext)
}

export function WhaleProvider (props: React.PropsWithChildren<any>): JSX.Element | null {
  const { network } = useNetworkContext()

  const client = useMemo(() => {
    return newWhaleClient(network)
  }, [network])

  return (
    <WhaleApiClientContext.Provider value={client}>
      {props.children}
    </WhaleApiClientContext.Provider>
  )
}

function newWhaleClient (network: EnvironmentNetwork): WhaleApiClient {
  const whaleOpts = { network: networkMapper(network) }
  switch (network) {
    case EnvironmentNetwork.MainNet:
      return new WhaleApiClient({ url: 'https://ocean.defichain.com', ...whaleOpts })
    case EnvironmentNetwork.TestNet:
      return new WhaleApiClient({ url: 'https://ocean.defichain.com', ...whaleOpts })
    case EnvironmentNetwork.RemotePlayground:
      return new WhaleApiClient({ url: 'https://playground.defichain.com', ...whaleOpts })
    case EnvironmentNetwork.LocalPlayground:
      return new WhaleApiClient({ url: 'http://localhost:19553', ...whaleOpts })
  }
}
