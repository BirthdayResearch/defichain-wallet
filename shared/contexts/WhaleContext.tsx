import { WhaleApiClient } from '@defichain/whale-api-client'
import React, { createContext, useContext, useMemo } from 'react'
import { EnvironmentNetwork } from '@environment'
import { useNetworkContext } from '@shared-contexts/NetworkContext'

const WhaleApiClientContext = createContext<WhaleApiClient>(undefined as any)

export function useWhaleApiClient (): WhaleApiClient {
  return useContext(WhaleApiClientContext)
}

export function WhaleProvider (props: React.PropsWithChildren<any>): JSX.Element | null {
  const { network } = useNetworkContext()

  const client = useMemo(() => {
    return newWhaleClient(network)
  }, [network])

  return (
    <WhaleApiClientContext.Provider value={client}>
      {props.children}
    </WhaleApiClientContext.Provider>
  )
}

function newWhaleClient (network: EnvironmentNetwork): WhaleApiClient {
  switch (network) {
    case EnvironmentNetwork.MainNet:
      return new WhaleApiClient({ url: 'https://ocean.defichain.com', network: 'mainnet', version: 'v0' })
    case EnvironmentNetwork.TestNet:
      return new WhaleApiClient({ url: 'https://ocean.defichain.com', network: 'testnet', version: 'v0' })
    case EnvironmentNetwork.RemotePlayground:
      return new WhaleApiClient({ url: 'https://playground.defichain.com', network: 'regtest', version: 'v0' })
    case EnvironmentNetwork.LocalPlayground:
      return new WhaleApiClient({ url: 'http://localhost:19553', network: 'regtest', version: 'v0' })
  }
}
