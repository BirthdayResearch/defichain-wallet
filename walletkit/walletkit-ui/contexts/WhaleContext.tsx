import { WhaleApiClient, WhaleRpcClient } from "@defichain/whale-api-client";
import {
  newOceanOptions,
  newWhaleAPIClient,
  newWhaleRpcClient,
} from "@waveshq/walletkit-core";
import React, { createContext, useContext, useMemo } from "react";

import { useNetworkContext } from "./NetworkContext";
import { useServiceProviderContext } from "./StoreServiceProvider";

const WhaleApiClientContext = createContext<{
  whaleAPI: WhaleApiClient;
  whaleRPC: WhaleRpcClient;
}>(undefined as any);

export function useWhaleApiClient(): WhaleApiClient {
  return useContext(WhaleApiClientContext).whaleAPI;
}

export function useWhaleRpcClient(): WhaleRpcClient {
  return useContext(WhaleApiClientContext).whaleRPC;
}

export function WhaleProvider({
  children,
}: React.PropsWithChildren<any>): JSX.Element | null {
  const { network } = useNetworkContext();

  // Do not destructure, some projects are not wrapped with ServiceProvider
  const service = useServiceProviderContext();

  const client = useMemo(
    () => ({
      whaleAPI: newWhaleAPIClient(newOceanOptions(network, service?.url)),
      whaleRPC: newWhaleRpcClient(newOceanOptions(network, service?.url)),
    }),
    [network, service?.url],
  );

  return (
    <WhaleApiClientContext.Provider value={client}>
      {children}
    </WhaleApiClientContext.Provider>
  );
}
