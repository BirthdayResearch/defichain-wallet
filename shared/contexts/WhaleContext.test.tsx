import { render } from '@testing-library/react-native'
import { Text, View } from 'react-native'
import { useWhaleApiClient, WhaleProvider } from './WhaleContext'
import * as NetworkContext from './NetworkContext'
import { EnvironmentNetwork } from '@environment'

const networkContext = jest.spyOn(NetworkContext, 'useNetworkContext')

describe('Whale Context test', () => {
  const networkDetails = [
    {
      url: 'https://ocean.defichain.com',
      network: 'mainnet',
      name: EnvironmentNetwork.MainNet
    },
    {
      url: 'https://ocean.defichain.com',
      network: 'testnet',
      name: EnvironmentNetwork.TestNet
    },
    {
      url: 'https://playground.defichain.com',
      network: 'regtest',
      name: EnvironmentNetwork.RemotePlayground
    },
    {
      url: 'http://localhost:19553',
      network: 'regtest',
      name: EnvironmentNetwork.LocalPlayground
    }
  ]

  networkDetails.forEach((networkDetail) => {
    describe(`Whale Context test for ${networkDetail.name}`, () => {
      it('should match snapshot', () => {
        networkContext.mockImplementation(() => ({ network: networkDetail.name }))
        function WhaleProviderComponent (): JSX.Element {
          const client = useWhaleApiClient()
          expect(Object.keys(client)).toEqual(
            expect.arrayContaining(
              ['options',
                'rpc',
                'address',
                'poolpairs',
                'transactions',
                'tokens',
                'masternodes',
                'blocks',
                'oracles',
                'prices',
                'stats',
                'rawtx',
                'fee',
                'loan'
              ]
            )
          )
          const options = {
            version: 'v0',
            timeout: 60000,
            url: networkDetail.url,
            network: networkDetail.network
          }
          expect(client.options).toMatchObject(options)
          return (
            <View>
              <Text>{JSON.stringify(client.options)}</Text>
            </View>
          )
        }

        const rendered = render(
          <WhaleProvider>
            <WhaleProviderComponent />
          </WhaleProvider>
        )
        expect(rendered.toJSON()).toMatchSnapshot()
      })
    })
  })
})
