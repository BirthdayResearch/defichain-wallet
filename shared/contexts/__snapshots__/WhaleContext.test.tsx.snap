// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Whale Context test Whale Context test for Local should match snapshot 1`] = `
<View>
  <Text>
    {"url":"http://localhost:19553","timeout":60000,"version":"v0","network":"regtest"}
  </Text>
</View>
`;

exports[`Whale Context test Whale Context test for MainNet should match snapshot 1`] = `
<View>
  <Text>
    {"url":"https://ocean.defichain.com","timeout":60000,"version":"v0","network":"mainnet"}
  </Text>
</View>
`;

exports[`Whale Context test Whale Context test for Playground should match snapshot 1`] = `
<View>
  <Text>
    {"url":"https://playground.defichain.com","timeout":60000,"version":"v0","network":"regtest"}
  </Text>
</View>
`;

exports[`Whale Context test Whale Context test for TestNet should match snapshot 1`] = `
<View>
  <Text>
    {"url":"https://ocean.defichain.com","timeout":60000,"version":"v0","network":"testnet"}
  </Text>
</View>
`;
